### PR TITLE
New Indoor positioning Sample

### PR DESCRIPTION
**Pull request by prupani-7**
## Description

PR to add a new Kotlin sample _"Show Device Location Using Indoor Positioning"_ in `Maps and Scenes` category.

## Links and Data

Sample Epic: `runtime/kotlin/issues/2011`

## What To Review

-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test

Replace line 107-108 from MainActivity.kt with this section of code, to test the sample using IPS webmap from redlands Building M. You need to be in Esri building M, to test this.

```
val itemUrl =
    "https://runtimecoretest.maps.arcgis.com/home/item.html?id=41fd5a7ee7044a40ad73f7f4a79a5a25"
val authenticator = ArcGISAuthenticationChallengeHandler {
    ArcGISAuthenticationChallengeResponse.ContinueWithCredential(
        TokenCredential.create(
            itemUrl,
            "username",
            "password",
            0
        ).getOrThrow()
    )
}
ArcGISEnvironment.authenticationManager.arcGISAuthenticationChallengeHandler = authenticator
mapView.map = ArcGISMap(itemUrl)
```

please see the linked kotlin issue for username and password

### CODE REVIEW

File: show-device-location-using-indoor-positioning/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+When there is no IPS beacons nearby, or other errors occur while initializing the indoors location data source, it will seamlessly fall back to the current device location as determined by GPS.
+
+## How it works
+
+1. Load an IPS-enabled map. This can be a web map hosted as a portal item in ArcGIS Online, an Enterprise Portal, or a mobile map package (.mmpk) created with ArcGIS Pro.
+2. Create an `IndoorsLocationDataSource` with the positioning feature table (stored with the map) and the pathways feature table after both tables are loaded.
+3. Handle location change events to respond to floor changes or read other metadata for locations.
+4. Assign the `IndoorsLocationDataSource` to the map view's location display.
+5. Enable and disable the map view's location display using `LocationDisplay.start()` and `LocationDisplay.stop()`. Device location will appear on the display as a blue dot and update as the user moves throughout the space.
+6. Use the `LocationDisplay.AutoPanMode` property to change how the map behaves when location updates are received.
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/README.md:24]:
```suggestion
6. Use the `LocationDisplay.setAutoPanMode()` function to change how the map behaves when location updates are received.
```

File: show-device-location-using-indoor-positioning/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+When there is no IPS beacons nearby, or other errors occur while initializing the indoors location data source, it will seamlessly fall back to the current device location as determined by GPS.
+
+## How it works
+
+1. Load an IPS-enabled map. This can be a web map hosted as a portal item in ArcGIS Online, an Enterprise Portal, or a mobile map package (.mmpk) created with ArcGIS Pro.
+2. Create an `IndoorsLocationDataSource` with the positioning feature table (stored with the map) and the pathways feature table after both tables are loaded.
+3. Handle location change events to respond to floor changes or read other metadata for locations.
+4. Assign the `IndoorsLocationDataSource` to the map view's location display.
+5. Enable and disable the map view's location display using `LocationDisplay.start()` and `LocationDisplay.stop()`. Device location will appear on the display as a blue dot and update as the user moves throughout the space.
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/README.md:23]: Looks like the api now is `LocationDisplay.dataSource.start()` and `LocationDisplay.dataSource.stop()`

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                orderByFields.add(
+                    OrderBy(
+                        dateCreatedFieldName,
+                        sortOrder = SortOrder.Descending
+                    )
+                )
+            }
+            lifecycleScope.launch {
+                positioningFeatureTable?.queryFeatures(queryParameters)
+                    ?.onSuccess { queryResults ->
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:172]:
```suggestion
                positioningFeatureTable.queryFeatures(queryParameters)
                    .onSuccess { queryResults ->
```

Looks like null checks are not needed along this chain.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+                        // perform search query using the queryParameters
+                        queryResults.forEach { feature ->
+                            // check if serviceFeatureTable contains positioning data
+                            // The ID that identifies a row in the positioning table.
+                            if (feature != null) {
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:177]:
```suggestion
```

`feature` is a non-null type as well.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+                                    getLevelsTable(),
+                                    positioningId
+                                )
+                                // start the location display (blue dot)
+                                startLocationDisplay()
+                            } else {
+                                showError("No features found")
+                            }
+                        }
+                    }?.onFailure {
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:199]:
```suggestion
                    }.onFailure {
```

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        lifecycleScope.launch {
+            locationDisplay.dataSource.start()
+        }
+
+        // coroutine scope to collect data source location changes like currentFloor, positionSource, transmitterCount, networkCount and horizontalAccuracy
+        lifecycleScope.launch {
+            locationDisplay.dataSource.locationChanged.collect { location ->
+                // get the location properties of the LocationDataSource
+                val locationProperties = location.additionalSourceProperties
+                if (locationProperties == null) {
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:273]:
```suggestion
```

`locationProperties` is a non-null type as well.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+                    LocationDataSourceStatus.FailedToStart -> {
+                        progressBar.visibility = View.GONE
+                        showError("Failed to start IndoorsLocationDataSource")
+                    }
+                    LocationDataSourceStatus.Stopped -> {
+                        progressBar.visibility = View.GONE
+                        mapView.locationDisplay.dataSource.stop()
+                        showError("IndoorsLocationDataSource stopped due to an internal error")
+                    }
+                    else -> null
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:327]:
```suggestion
                    else -> {}
```

I think it's okay to have a empty block instead of null to avoid the warning.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        return try {
+            val levelsFeatureLayer =
+                mapView.map?.operationalLayers?.firstOrNull { it.name == "levels" } as? FeatureLayer
+            levelsFeatureLayer?.featureTable as? ArcGISFeatureTable
+        } catch (e: Exception) {
+            // if levels table not found in map's operationalLayers
+            showError("Levels table not found")
+            null
+        }
+    }
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:251]: The long null check chains in these functions are kinda hard to follow. You could rescope it with a `let` and `?:` with a `run` block if you a message has to be shown. Also I think the `getPathwaysTable()` and `getLevelsTable()` could be made into a single function that takes the layer `name` as the parameter.

```suggestion
    private fun getFeatureTable(name: String): ArcGISFeatureTable? {
        return mapView.map?.operationalLayers?.firstOrNull { operationalLayer ->
            operationalLayer.name == name
        }?.let { layer -> 
            if (layer is FeatureLayer) {
                layer.featureTable as ArcGISFeatureTable
            } else {
                null
            }
        } ?: run {
            showError("$name table not found")
            null
        }
    }
```


Something like this should work.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                val transmitterCount = (locationProperties["transmitterCount"] ?: "").toString()
+                val satelliteCount = (locationProperties["satelliteCount"] ?: "").toString()
+
+                // check if current floor hasn't been set or if the floor has changed
+                val newFloor = floor.toInt()
+                if (currentFloor == null || currentFloor != newFloor) {
+                    currentFloor = newFloor
+                    // update layer's definition express with the current floor
+                    val layerList = mapView.map?.operationalLayers
+                    layerList?.forEach { layer ->
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:288]:
```suggestion
                    mapView.map?.operationalLayers?.forEach { layer ->
```
This could be simplified since we're not using the `layerList` variable.

**Review by kaushikrw — CHANGES_REQUESTED** [review]
@prupani-7 I have yet to test the app on campus but looks like with the recent commit there are conflicts when I pull the branch. Could you check on that please? You might also want to merge in `v.next`. I also noticed that because of recent api changes some properties are no longer nullable. I have marked some of them as well.

File: show-device-location-using-indoor-positioning/README.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
+# Show device location using indoor positioning
+
+Show your device's real-time location while inside a building by using signals from indoor positioning beacons.
```
Comment (prupani-7) [inline show-device-location-using-indoor-positioning/README.md:3]: Show your device's real-time location while inside a building by using signals from indoor positioning beacon.

Comment (prupani-7) [inline show-device-location-using-indoor-positioning/README.md:3]:
```suggestion
Show your device's real-time location while inside a building by using signals from indoor positioning beacon.
```

**Comment by prupani-7** [comment]
@kaushikrw Thanks for the review. I addressed your changes!

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+                            // Setting up IndoorsLocationDataSource with positioning, pathways tables and positioning ID.
+                            // positioningTable - the "ips_positioning" feature table from an IPS-enabled map.
+                            // pathwaysTable - An ArcGISFeatureTable that contains pathways as per the ArcGIS Indoors Information Model.
+                            // levelsTable - An ArcGISFeatureTable that contains floor levels in accordance with the ArcGIS Indoors Information Model.
+                            // Setting this property enables path snapping of locations provided by the IndoorsLocationDataSource.
+                            // positioningID - an ID which identifies a specific row in the positioningTable that should be used for setting up IPS.
+                            mIndoorsLocationDataSource = IndoorsLocationDataSource(
+                                positioningFeatureTable,
+                                getFeatureTable("Pathways"),
+                                getFeatureTable("levels"),
```
Comment (kaushikrw) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:198]:
```suggestion
                                getFeatureTable("Levels"),
```

This must be case sensitive I think? I was seeing an error - `levels table not found`

**Review by kaushikrw — CHANGES_REQUESTED** [review]
@prupani-7 Thanks. It looks good, just some doc related comments. 

I did notice while testing that, if there is no IPS signal the fallback to GPS was not visible like the readme file says. Is there a specific criteria for when that shows up?

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 11 above)
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:198]: I had it as `levels` before updating this code, and it was working, in the way, it updated the levels as I moved up and down the floors. Did you test this today, with the updated code?

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+                ignoreCase = true
+            ) || it.name.equals("Date_Created", ignoreCase = true)
+        }
+        return field?.name
+    }
+
+    /**
+     * Retrieves the "Pathways" or the "levels" table depending on the string passed.
+     * Pathways table is an [ArcGISFeatureTable] that contains pathways as per the ArcGIS Indoors Information Model. Setting this property enables path snapping of locations provided by the IndoorsLocationDataSource.
+     * Levels tables which is an [ArcGISFeatureTable] that contains floor levels in accordance with the ArcGIS Indoors Information Model. Providing this table enables the retrieval of a location's floor level ID.
```
[the remark this answers is not in the document]
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:229]: actually these comments are redundant. I noticed talk about them when the `IndoorsLocationDataSource` object is created. So I am going to remove these from here.

**Comment by prupani-7** [comment]
> I did notice while testing that, if there is no IPS signal the fallback to GPS was not visible like the readme file says. Is there a specific criteria for when that shows up?

It does fallback to GPS, as verified with you on the call.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 11 above)
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: Updated

**Comment by prupani-7** [comment]
Tested with android 13 device, with the new permissions added `BLUETOOTH_SCAN`[ See slack conversation ](https://esri-runtime.slack.com/archives/C04U8JCA63W/p1679525098338349)
image

**Comment by prupani-7** [comment]
@kaushikrw could you re-review this PR?

**Comment by kaushikrw** [comment]
Looks good to me.

**Review by kaushikrw — APPROVED** [review]

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+            val positioningFeatureTable = featureTable as ServiceFeatureTable
+            // when multiple entries are available, IndoorsLocationDataSource constructor function
+            // looks up the entry with the most recent date and takes this positioning data
+            // set up queryParameters to grab one result.
+            val dateCreatedFieldName =
+                getDateCreatedFieldName(positioningFeatureTable.fields)
+            if (dateCreatedFieldName == null) {
+                showError("The service table does not contain \"DateCreated\" fields.")
+                return
+            }
```
Comment (shubham7109) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:176]: We could use the Elvis operator to condense this down to 2 lines. 

```suggestion
            val dateCreatedFieldName = getDateCreatedFieldName(positioningFeatureTable.fields)
                ?: return showError("The service table does not contain \"DateCreated\" fields.")
```

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+        }
+    }
+
+    /**
+     * Sets up the [mIndoorsLocationDataSource] using the IPS_Positioning table
+     */
+    private fun setupIndoorsLocationDataSource(featureTable: FeatureTable) {
+        // positioningTable needs to be present
+        if (featureTable.tableName == "IPS_Positioning") {
```
Comment (shubham7109) [inline show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:166]: Can this check be done in the caller function? 
```kotlin
if (featureTable.tableName == "IPS_Positioning") {
    setupIndoorsLocationDataSource(featureTable)
}
```

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+                                feature.attributes[positioningFeatureTable.globalIdField].toString()
+                            val positioningId = Guid(globalID)
+                            // Setting up IndoorsLocationDataSource with positioning, pathways tables and positioning ID.
+                            // positioningTable - the "ips_positioning" feature table from an IPS-enabled map.
+                            // pathwaysTable - An ArcGISFeatureTable that contains pathways as per the ArcGIS Indoors Information Model.
+                            // Setting this property enables path snapping of locations provided by the IndoorsLocationDataSource.
+                            // levelsTable - An ArcGISFeatureTable that contains floor levels in accordance with the ArcGIS Indoors Information Model.
+                            // Providing this table enables the retrieval of a location's floor level ID.
+                            // positioningID - an ID which identifies a specific row in the positioningTable that should be used for setting up IPS.
+                            mIndoorsLocationDataSource = IndoorsLocationDataSource(
```
[the remark this answers is not in the document]
Comment (shubham7109) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:207]: Yes, that seems to be the case since the `maxFeatures` is set to 1, so the `mIndoorsLocationDataSource` will be set once. But the code doesn't have to be looping through either, so could be updated with: 

```kotlin
.onSuccess { queryResults ->
                        val featureResult = queryResults.first()
```

**Review by shubham7109 — CHANGES_REQUESTED** [review]
Nice sample @prupani-7. I've requested a few changes here.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+                            // start the location display (blue dot)
+                            startLocationDisplay()
+                        }
+                    }.onFailure {
+                        showError("The positioning table contain no data")
+                    }
+            }
+        } else {
+            showError("Positioning Table not found in FeatureTables")
+        }
```
[the remark this answers is not in the document]
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt:222]: makes sense, but everything else like the creation of queryparameters and the indoorsLocationDataSource object, still needs to be under the if (featureTable.tableName == "IPS_Positioning") condition as there could be other feature tables as well.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 14 above)
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: I modified it as below
if (featureTable.tableName != "IPS_Positioning") {
            showError("Positioning Table not found in FeatureTables")
            return
        } // IPS_Positioning table needs to be present
        else if (featureTable.tableName == "IPS_Positioning") {
        ....
        }
        
I think either approach should be fine.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: Updated it.

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+                        DecimalFormat(".##").format(
+                            it.horizontalAccuracy
+                        )
+                    } + "m, "
+                if (positionSource == Location.SourceProperties.Values.POSITION_SOURCE_GNSS) {
+                    locationPropertiesMessage += "Satellite-count: $satelliteCount"
+                } else if (positionSource == "BLE") {
+                    locationPropertiesMessage += "Transmitter-count: $transmitterCount"
+                }
+                textView.text = locationPropertiesMessage
```
[the remark this answers is not in the document]
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: updated

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 14 above)
Comment (shubham7109) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: The `else` condition is not nesseary here since if `featureTable.tableName != "IPS_Positioning"` is true the function will return.

**Review by shubham7109 — APPROVED** [review]

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (fpearson1111) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: I don't think an extra condition is needed because if `featureTable.tableName != "IPS_Positioning"` is false then the table name must be `IPS_Positioning`, so there's no need to add an extra check for `featureTable.tableName == "IPS_Positioning"` because it will always be true.

Either way I think it's better to go with the suggestion in https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/80/files/836e3d07f4684c8f6f938b8b2635784d68ef7373#r1146930505 instead.

**Review by fpearson1111 — APPROVED** [review]
One more piece of feedback but happy to approve, LGTM

File: show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt (hunk 14 above)
Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]: Done! :)

Comment (prupani-7) [reply show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt]:
```
// IPS_Positioning table needs to be present
                    if (featureTable.tableName == "IPS_Positioning") {
                        setupIndoorsLocationDataSource(featureTable)
                    } else {
                        showError("Positioning Table not found in FeatureTables")
                    }
```
